rfcs/0136-stabilize-incrementally.md

```diff
@@ -0,0 +1,386 @@
+
+   - Afterwards, Flakes itself and its CLI components can be stabilized. The final design of Flakes will also require another RFC.
+
+# Problem statement
+[problem-statement]: #problem-statement
+
+For the past few years, we've accumulated a massive backlog of unstable features in the form of the new command-line interface and flakes.
+There is now a growing desire to stabilize those features, but as of yet no plan on exactly how to do so.
+The tension between that desire and a lack of a clear plan has loomed over our heads for a while; this RFC aims to dispel it by providing a concrete plan, a plan that hopefully will mitigate the lingering controversies and tensions around Flakes.
+
+## The situation so far
+
+There are a few facts that must be introduced to give context to this situation.
+We can't change these short of a time machine, so we must find a way to live with them instead.
+
+### Flakes are very popular
+
+As measured in the community polls, Nix has a lot of new users, dwarfing the number of long-time users.
+Flakes is very popular among these new users.
+Ergo, Flakes is very popular among the Nix community as a whole.
+
+Many groups and individuals interested in the continued growth of the Nix community see that Flakes are popular, and wish for them to be stabilized to attract still more users.
+The thinking is that if unstable Flakes are already proven to be popular, stable Flakes will be even moreso.
+
+### Difficulties in the roll-out
+
+It is undeniable that the roll-out of these new features has been controversial.
+Graham Christensen's blog post [flakes-are-an-obviously-good-thing](https://grahamc.com/blog/flakes-are-an-obviously-good-thing) lays out the procedure issues along the way quite well.
+
+Some people were upset the Flakes RFC was abandoned but the feature merged.
+Other people were fine with experimental features being merged without RFC, but were upset because Flakes never really *felt* experimental.
+Experimental features are expected to be subject to community feedback, modified a lot based upon that feedback, and, most importantly, *discouraged* from being used in production.
+
+### Flakes are criticized for encroaching on other features
+
+There are many criticisms about Flakes.
+But one of them especially relevant to stabilizing is a perception that Flakes have encroached on other new features, in the sense that it ought to be possible to use those other features without Flakes but isn't in practice.
+For example, there is no reason in theory that pure evaluation of Nix expressions requires Flakes.
+But without the ability to populate some sort of initial list of store paths that are safe to import, pure evaluation does in practice require Flakes.
+
+This is especially noticeable for new CLI features that *previously did*, in fact, work without Flakes.
+For example, in earlier versions of Nix, `nix search` worked without Flakes.
```

Comment:
This is my main issue with the current CLI state. The RFC text further down kind of hints at that this will be fixed, but it is too vague to be certain.

Explicitly mentioning `nix search` as an example would make me a lot more in favor of this RC ;)

Reply:
Another is the error message you get from `nix run` on an expression that isn't a derivation:

```
$ nix eval -f . libcIconv
<LAMBDA>
$ nix run -f . libcIconv
error: attribute 'type' does not exist
```

A lot of the code paths that can be used for both flakes and non-flakes appear to have had the error handling removed from the non-flake case and replaced with "`goto assume-it-is-a-flake`".

Reply:
@piegamesde so I/we purposely didn't want to state that stabilizing is blocked on developing new nin-flake eatures that probably wouldn't inform the final design of flakes. Even though those things used to work pre flakes given the state of the code today it is a new feature.

However, on a personal level, I do help it will be done first.

Reply:
IMO, the stabilization of flakes should not happen without having at least feature parity with the old CLI when flakes are disabled. I don't agree with framing that as a "new feature".

Personally, I'd additionally like to have that *all* CLI features that get stabilized must also work with flakes disabled. But I understand if that's too much to ask for.

Reply:
That would be okay by me, but let's not extend the scope of the PR further for now, it's a net positive as is.